bins/stackabled/src/middleware/authentication.rs

```diff
@@ -0,0 +1,173 @@
+pub enum HtpasswdError {
+    #[snafu(display("failed to read htpasswd file at {path:?}"))]
+    Read {
+        source: std::io::Error,
+        path: PathBuf,
+    },
+    #[snafu(display("malformed htpasswd entry on line {line}"))]
+    Entry {
+        source: HtpasswdEntryError,
+        line: usize,
+    },
+}
+#[derive(Debug, Snafu)]
+#[snafu(module)]
+pub enum HtpasswdEntryError {
+    #[snafu(display("invalid hash type (only bcrypt is currently supported)"))]
+    InvalidHashType,
+    #[snafu(display("no username/password separator"))]
+    NoSeparator,
+}
+
+impl Authenticator {
+    pub fn load_htpasswd(path: &Path) -> Result<Self, HtpasswdError> {
+        use htpasswd_entry_error::*;
+        use htpasswd_error::*;
+        let htaccess = std::fs::read_to_string(path).context(ReadSnafu { path })?;
+        let mut accounts = HashMap::new();
+        for (line, entry) in htaccess.lines().enumerate() {
+            if let Some((username, prefixed_pw_hash)) = entry.split_once(':') {
+                if prefixed_pw_hash.starts_with("$2y$") {
+                    accounts.insert(
+                        Username(username.to_string()),
+                        PasswordHash::Bcrypt(prefixed_pw_hash.to_string()),
+                    );
+                    Ok(())
+                } else {
+                    InvalidHashTypeSnafu.fail()
+                }
+            } else {
+                NoSeparatorSnafu.fail()
+            }
+            .context(EntrySnafu { line })?
+        }
+        Ok(Self {
+            state: Arc::new(State {
+                accounts,
+                sessions: RwLock::new(HashMap::new()),
+            }),
+        })
+    }
+
+    fn validate_login(&self, headers: &HeaderMap<HeaderValue>) -> Result<Username, &'static str> {
+        if let Some(session_token) = headers.typed_get::<axum::headers::Authorization<Bearer>>() {
+            if let Some(username) = self
+                .state
+                .sessions
+                .read()
+                .unwrap()
+                .get(session_token.0.token())
+            {
+                Ok(username.clone())
+            } else {
+                Err("request authentication failed: invalid session token")
+            }
+        } else if let Some(basic_creds) = headers.typed_get::<axum::headers::Authorization<Basic>>()
+        {
+            let username = basic_creds.username();
+            let password_hash = self.state.accounts.get(username);
+            if password_hash
+                .and_then(|PasswordHash::Bcrypt(hash)| {
+                    bcrypt::verify(basic_creds.password(), hash).ok()
+                })
+                .unwrap_or(false)
+            {
+                Ok(Username(username.to_string()))
+            } else {
+                Err("request authentication failed: invalid username or password")
+            }
+        } else {
+            Err("request authentication failed: no valid authentication method presented")
```

Comment:
Should we include the supported methods in the error message?

Reply:
a6e8a8e626a65759749fa71d448a1fc5b36622d5